chainer/utils/meta.py

```diff
@@ -0,0 +1,79 @@
+import warnings
+
+import six
+
+
+def final(*args, **kwargs):
+    """Decorator to declare a method final.
+
+    By default, :class:`TypeError` is raised when the decorated method is being
+    overridden.
+
+    The class in which the decorated method is defined must inherit from a
+    base class returned by :meth:`~chainer.utils.enable_final`.
+
+    Args:
+        action: Specifies what happens when the decorated method is being
+            overridden. If `'error'`, :class:`TypeError` is raised. If
+            :class:`DeprecationWarning`, a deprecation warning is raised.
+    """
+    def wrap(f, action='error'):
+        assert callable(f)
+        f.__is_final = (action,)
+        return f
+
+    # apply directly
+    if not kwargs and len(args) == 1:
+        f, = args
+        return wrap(f)
+
+    # with arguments
+    assert len(args) == 0
+
+    def w(f):
+        return wrap(f, **kwargs)
+
+    return w
+
+
+class _EnableFinal(type):
+    def __new__(cls, name, bases, d):
+        for k in d:
+            for base in bases:
+                f = getattr(base, k, None)  # base method
+                if hasattr(f, '__is_final'):
+                    action, = getattr(f, '__is_final')
```

Comment:
Just write `f.__is_final`.

Reply:
It's not possible. It would be regarded as `f._EnableFinal__is_final`.

Reply:
Oh, ok.